tests/dt.go

```diff
@@ -0,0 +1,234 @@
+	nodeMap := map[string]corev1.Node{}
+	nodes, err := oa.kubeCli.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes.Items {
+		nodeMap[node.Name] = node
+	}
+
+	pds, err := oa.kubeCli.CoreV1().Pods(cluster.Namespace).List(
+		metav1.ListOptions{LabelSelector: labels.SelectorFromSet(
+			label.New().Instance(cluster.ClusterName).PD().Labels(),
+		).String()})
+	if err != nil {
+		return err
+	}
+	err = oa.checkDisasterTolerance(pds.Items, nodeMap)
+	if err != nil {
+		return err
+	}
+
+	tikvs, err := oa.kubeCli.CoreV1().Pods(cluster.Namespace).List(
+		metav1.ListOptions{LabelSelector: labels.SelectorFromSet(
+			label.New().Instance(cluster.ClusterName).TiKV().Labels(),
+		).String()})
+	if err != nil {
+		return err
+	}
+	err = oa.checkDisasterTolerance(tikvs.Items, nodeMap)
+	if err != nil {
+		return err
+	}
+
+	tidbs, err := oa.kubeCli.CoreV1().Pods(cluster.Namespace).List(
+		metav1.ListOptions{LabelSelector: labels.SelectorFromSet(
+			label.New().Instance(cluster.ClusterName).TiDB().Labels(),
+		).String()})
+	if err != nil {
+		return err
+	}
+	return oa.checkDisasterTolerance(tidbs.Items, nodeMap)
+}
+
+func (oa *operatorActions) checkDisasterTolerance(allPods []corev1.Pod, nodeMap map[string]corev1.Node) error {
+	rackPods := map[string][]corev1.Pod{}
+	for _, pod := range allPods {
+		if node, exist := nodeMap[pod.Spec.NodeName]; exist {
+			pods, exist := rackPods[node.Labels[RackLabel]]
+			if !exist {
+				pods = []corev1.Pod{}
+			}
+			pods = append(pods, pod)
+			rackPods[node.Labels[RackLabel]] = pods
+		}
+	}
+
+	podNum := len(allPods)
+	maxPodsOneRack := podNum / RackNum
+	mod := podNum % RackNum
+	if mod > 0 {
+		maxPodsOneRack = maxPodsOneRack + 1
+	}
+
+	for rack, pods := range rackPods {
+		if len(pods) > maxPodsOneRack {
+			return fmt.Errorf("the rack:[%s] have pods more than %d", rack, maxPodsOneRack)
+		}
+	}
+	return nil
+}
+
+func (oa *operatorActions) CheckDisasterToleranceOrDie(cluster *TidbClusterConfig) {
+	err := oa.CheckDisasterTolerance(cluster)
+	if err != nil {
+		slack.NotifyAndPanic(err)
+	}
+}
+
+func (oa *operatorActions) CheckDataRegionDisasterToleranceOrDie(cluster *TidbClusterConfig) {
+	err := oa.CheckDataRegionDisasterTolerance(cluster)
+	if err != nil {
+		slack.NotifyAndPanic(err)
+	}
+}
+
+func (oa *operatorActions) CheckDataRegionDisasterTolerance(cluster *TidbClusterConfig) error {
+	pdClient := http.Client{
+		Timeout: 10 * time.Second,
+	}
+	url := fmt.Sprintf("http://%s-pd.%s:2379/pd/api/v1/regions", cluster.ClusterName, cluster.Namespace)
+	resp, err := pdClient.Get(url)
+	if err != nil {
+		return err
+	}
+	buf, _ := ioutil.ReadAll(resp.Body)
+	regions := &RegionsInfo{}
+	err = json.Unmarshal(buf, &regions)
+	if err != nil {
+		return err
+	}
+
+	rackNodeMap, err := oa.getNodeRackMap()
+	if err != nil {
+		return err
+	}
+
+	for _, region := range regions.Regions {
```

Comment:
Could you please add comment for this? I find this is hard to understand at first glance